ClassCastException when with @EmebedddedId and @OneToOne relationship

See https://github.com/quarkusio/quarkus/issues/47536#issuecomment-2838802863

```java
[ERROR] org.acme.hibernate.orm.panache.ReproTest.test(TransactionalUniAsserter) -- Time elapsed: 0.124 s <<< ERROR!
java.lang.ClassCastException: class org.hibernate.sql.results.graph.entity.internal.EntitySelectFetchInitializer cannot be cast to class org.hibernate.reactive.sql.results.graph.ReactiveInitializer (org.hibernate.sql.results.graph.entity.internal.EntitySelectFetchInitializer and org.hibernate.reactive.sql.results.graph.ReactiveInitializer are in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @4cc7d00d)
	at org.hibernate.reactive.sql.results.graph.embeddable.internal.ReactiveEmbeddableInitializerImpl.lambda$forEachReactiveSubInitializer$0(ReactiveEmbeddableInitializerImpl.java:75)
	at org.hibernate.reactive.util.impl.CompletionStages.lambda$loop$2(CompletionStages.java:175)
	at org.hibernate.reactive.util.impl.CompletionStages.lambda$loop$7(CompletionStages.java:410)
	at org.hibernate.reactive.util.impl.CompletionStages$ArrayLoop.next(CompletionStages.java:484)
```

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/metamodel/mapping/internal/ReactiveEmbeddedIdentifierMappingImpl.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/metamodel/mapping/internal/ReactiveEmbeddedIdentifierMappingImpl.java
@@ -13,6 +13,7 @@ import org.hibernate.metamodel.internal.AbstractCompositeIdentifierMapping;
 import org.hibernate.metamodel.mapping.EmbeddableMappingType;
 import org.hibernate.metamodel.mapping.JdbcMapping;
 import org.hibernate.metamodel.mapping.internal.EmbeddedIdentifierMappingImpl;
+import org.hibernate.metamodel.mapping.internal.ToOneAttributeMapping;
 import org.hibernate.reactive.sql.results.graph.embeddable.internal.ReactiveEmbeddableFetchImpl;
 import org.hibernate.spi.NavigablePath;
 import org.hibernate.sql.ast.spi.SqlSelection;
@@ -20,6 +21,7 @@ import org.hibernate.sql.ast.tree.from.TableGroup;
 import org.hibernate.sql.results.graph.DomainResultCreationState;
 import org.hibernate.sql.results.graph.Fetch;
 import org.hibernate.sql.results.graph.FetchParent;
+import org.hibernate.sql.results.graph.Fetchable;
 
 public class ReactiveEmbeddedIdentifierMappingImpl extends AbstractCompositeIdentifierMapping {
 
@@ -103,5 +105,14 @@ public class ReactiveEmbeddedIdentifierMappingImpl extends AbstractCompositeIden
 	@Override
 	public String getFetchableName() {
 		return delegate.getFetchableName();
+	}
+
+	@Override
+	public Fetchable getFetchable(int position) {
+		Fetchable fetchable = delegate.getFetchable( position );
+		if ( fetchable instanceof ToOneAttributeMapping ) {
+			return new ReactiveToOneAttributeMapping( (ToOneAttributeMapping) fetchable );
+		}
+		return fetchable;
 	}
 }

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/sql/results/graph/embeddable/internal/ReactiveEmbeddableAssembler.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/sql/results/graph/embeddable/internal/ReactiveEmbeddableAssembler.java
@@ -1,0 +1,42 @@
+/* Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright: Red Hat Inc. and Hibernate Authors
+ */
+package org.hibernate.reactive.sql.results.graph.embeddable.internal;
+
+import org.hibernate.reactive.sql.exec.spi.ReactiveRowProcessingState;
+import org.hibernate.reactive.sql.results.graph.ReactiveDomainResultsAssembler;
+import org.hibernate.reactive.sql.results.graph.ReactiveInitializer;
+import org.hibernate.sql.results.graph.Initializer;
+import org.hibernate.sql.results.graph.InitializerData;
+import org.hibernate.sql.results.graph.embeddable.EmbeddableInitializer;
+import org.hibernate.sql.results.graph.embeddable.internal.EmbeddableAssembler;
+import org.hibernate.sql.results.jdbc.spi.JdbcValuesSourceProcessingOptions;
+
+import java.util.concurrent.CompletionStage;
+
+import static org.hibernate.reactive.util.impl.CompletionStages.completedFuture;
+
+/**
+ * @see org.hibernate.sql.results.graph.embeddable.internal.EmbeddableAssembler
+ */
+public class ReactiveEmbeddableAssembler extends EmbeddableAssembler implements ReactiveDomainResultsAssembler {
+
+	public ReactiveEmbeddableAssembler(EmbeddableInitializer<?> initializer) {
+		super( initializer );
+	}
+
+	@Override
+	public CompletionStage<Object> reactiveAssemble(ReactiveRowProcessingState rowProcessingState, JdbcValuesSourceProcessingOptions options) {
+		final ReactiveInitializer<InitializerData> reactiveInitializer = (ReactiveInitializer<InitializerData>) getInitializer();
+		final InitializerData data = reactiveInitializer.getData( rowProcessingState );
+		final Initializer.State state = data.getState();
+		if ( state == Initializer.State.KEY_RESOLVED ) {
+			return reactiveInitializer
+					.reactiveResolveInstance( data )
+					.thenApply( v -> reactiveInitializer.getResolvedInstance( data ) );
+		}
+		return completedFuture( reactiveInitializer.getResolvedInstance( data ) );
+	}
+}

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/sql/results/graph/embeddable/internal/ReactiveEmbeddableFetchImpl.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/sql/results/graph/embeddable/internal/ReactiveEmbeddableFetchImpl.java
@@ -6,14 +6,20 @@
 package org.hibernate.reactive.sql.results.graph.embeddable.internal;
 
 import org.hibernate.engine.FetchTiming;
+import org.hibernate.reactive.sql.results.graph.entity.internal.ReactiveEntityFetchSelectImpl;
 import org.hibernate.spi.NavigablePath;
 import org.hibernate.sql.results.graph.AssemblerCreationState;
+import org.hibernate.sql.results.graph.DomainResultAssembler;
 import org.hibernate.sql.results.graph.DomainResultCreationState;
+import org.hibernate.sql.results.graph.Fetch;
 import org.hibernate.sql.results.graph.FetchParent;
+import org.hibernate.sql.results.graph.Fetchable;
+import org.hibernate.sql.results.graph.Initializer;
 import org.hibernate.sql.results.graph.InitializerParent;
 import org.hibernate.sql.results.graph.embeddable.EmbeddableInitializer;
 import org.hibernate.sql.results.graph.embeddable.EmbeddableValuedFetchable;
 import org.hibernate.sql.results.graph.embeddable.internal.EmbeddableFetchImpl;
+import org.hibernate.sql.results.graph.entity.internal.EntityFetchSelectImpl;
 
 public class ReactiveEmbeddableFetchImpl extends EmbeddableFetchImpl {
 
@@ -36,5 +42,21 @@ public class ReactiveEmbeddableFetchImpl extends EmbeddableFetchImpl {
 			InitializerParent<?> parent,
 			AssemblerCreationState creationState) {
 		return new ReactiveEmbeddableInitializerImpl( this, getDiscriminatorFetch(), parent, creationState, true );
+	}
+
+	@Override
+	public DomainResultAssembler<?> createAssembler(InitializerParent<?> parent, AssemblerCreationState creationState) {
+		Initializer<?> initializer = creationState.resolveInitializer( this, parent, this );
+		EmbeddableInitializer<?> embeddableInitializer = initializer.asEmbeddableInitializer();
+		return new ReactiveEmbeddableAssembler( embeddableInitializer );
+	}
+
+	@Override
+	public Fetch findFetch(Fetchable fetchable) {
+		Fetch fetch = super.findFetch( fetchable );
+		if ( fetch instanceof EntityFetchSelectImpl entityFetchSelect ) {
+			return new ReactiveEntityFetchSelectImpl( entityFetchSelect );
+		}
+		return fetch;
 	}
 }

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/sql/results/graph/embeddable/internal/ReactiveEmbeddableInitializerImpl.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/sql/results/graph/embeddable/internal/ReactiveEmbeddableInitializerImpl.java
@@ -5,12 +5,18 @@
  */
 package org.hibernate.reactive.sql.results.graph.embeddable.internal;
 
+
 import java.util.concurrent.CompletionStage;
 import java.util.function.BiFunction;
 
 import org.hibernate.metamodel.mapping.EmbeddableMappingType;
+import org.hibernate.metamodel.mapping.VirtualModelPart;
+import org.hibernate.metamodel.spi.EmbeddableInstantiator;
+import org.hibernate.reactive.sql.exec.spi.ReactiveRowProcessingState;
+import org.hibernate.reactive.sql.results.graph.ReactiveDomainResultsAssembler;
 import org.hibernate.reactive.sql.results.graph.ReactiveInitializer;
 import org.hibernate.sql.results.graph.AssemblerCreationState;
+import org.hibernate.sql.results.graph.DomainResultAssembler;
 import org.hibernate.sql.results.graph.Initializer;
 import org.hibernate.sql.results.graph.InitializerData;
 import org.hibernate.sql.results.graph.InitializerParent;
@@ -19,8 +25,13 @@ import org.hibernate.sql.results.graph.embeddable.EmbeddableResultGraphNode;
 import org.hibernate.sql.results.graph.embeddable.internal.EmbeddableInitializerImpl;
 import org.hibernate.sql.results.jdbc.spi.RowProcessingState;
 
+import static org.hibernate.reactive.util.impl.CompletionStages.completedFuture;
 import static org.hibernate.reactive.util.impl.CompletionStages.loop;
+import static org.hibernate.reactive.util.impl.CompletionStages.nullFuture;
 import static org.hibernate.reactive.util.impl.CompletionStages.voidFuture;
+import static org.hibernate.reactive.util.impl.CompletionStages.whileLoop;
+import static org.hibernate.sql.results.graph.embeddable.EmbeddableLoadingLogger.EMBEDDED_LOAD_LOGGER;
+import static org.hibernate.sql.results.graph.entity.internal.BatchEntityInsideEmbeddableSelectFetchInitializer.BATCH_PROPERTY;
 
 public class ReactiveEmbeddableInitializerImpl extends EmbeddableInitializerImpl
 		implements ReactiveInitializer<EmbeddableInitializerImpl.EmbeddableInitializerData> {
@@ -31,6 +42,10 @@ public class ReactiveEmbeddableInitializerImpl extends EmbeddableInitializerImpl
 				EmbeddableInitializerImpl initializer,
 				RowProcessingState rowProcessingState) {
 			super( initializer, rowProcessingState );
+		}
+
+		public Object[] getRowState(){
+			return rowState;
 		}
 
 		@Override
@@ -64,8 +79,126 @@ public class ReactiveEmbeddableInitializerImpl extends EmbeddableInitializerImpl
 
 	@Override
 	public CompletionStage<Void> reactiveResolveInstance(EmbeddableInitializerData data) {
-		super.resolveInstance( data );
+		if ( data.getState() != State.KEY_RESOLVED ) {
+			return voidFuture();
+		}
+
+		data.setState( State.RESOLVED );
+		return extractRowState( (ReactiveEmbeddableInitializerData) data )
+				.thenAccept( unused -> prepareCompositeInstance( (ReactiveEmbeddableInitializerData) data ) );
+	}
+
+	private CompletionStage<Void> extractRowState(ReactiveEmbeddableInitializerData data) {
+		final DomainResultAssembler<?>[] subAssemblers = assemblers[data.getSubclassId()];
+		final RowProcessingState rowProcessingState = data.getRowProcessingState();
+		final Object[] rowState = data.getRowState();
+		final boolean[] stateAllNull = {true};
+		final int[] index = {0};
+		final boolean[] forceExit = { false };
+		return whileLoop(
+				() -> index[0] < subAssemblers.length && !forceExit[0],
+				() -> {
+					final int i = index[0]++;
+					final DomainResultAssembler<?> assembler = subAssemblers[i];
+					if ( assembler instanceof ReactiveDomainResultsAssembler<?> reactiveAssembler ) {
+						return reactiveAssembler.reactiveAssemble( (ReactiveRowProcessingState) rowProcessingState )
+								.thenAccept( contributorValue -> setContributorValue(
+										contributorValue,
+										i,
+										rowState,
+										stateAllNull,
+										forceExit
+								) );
+					}
+					else {
+						setContributorValue(
+								assembler == null ? null : assembler.assemble( rowProcessingState ),
+								i,
+								rowState,
+								stateAllNull,
+								forceExit
+						);
+						return voidFuture();
+					}
+		})
+				.whenComplete(
+						(unused, throwable) -> {
+							if ( stateAllNull[0] ) {
+								data.setState( State.MISSING );
+							}
+						}
+				);
+	}
+
+	private void setContributorValue(
+			Object contributorValue,
+			int index,
+			Object[] rowState,
+			boolean[] stateAllNull,
+			boolean[] forceExit) {
+		if ( contributorValue == BATCH_PROPERTY ) {
+			rowState[index] = null;
+		}
+		else {
+			rowState[index] = contributorValue;
+		}
+		if ( contributorValue != null ) {
+			stateAllNull[0] = false;
+		}
+		else if ( isPartOfKey() ) {
+			// If this is a foreign key and there is a null part, the whole thing has to be turned into null
+			stateAllNull[0] = true;
+			forceExit[0] = true;
+		}
+	}
+
+	private CompletionStage<Void> prepareCompositeInstance(ReactiveEmbeddableInitializerData data) {
+		// Virtual model parts use the owning entity as container which the fetch parent access provides.
+		// For an identifier or foreign key this is called during the resolveKey phase of the fetch parent,
+		// so we can't use the fetch parent access in that case.
+		final ReactiveInitializer<ReactiveEmbeddableInitializerData> parent = (ReactiveInitializer<ReactiveEmbeddableInitializerData>) getParent();
+		if ( parent != null && getInitializedPart() instanceof VirtualModelPart && !isPartOfKey() && data.getState() != State.MISSING ) {
+			final ReactiveEmbeddableInitializerData subData = parent.getData( data.getRowProcessingState() );
+			return parent
+					.reactiveResolveInstance( subData )
+					.thenCompose(
+							unused -> {
+								data.setInstance( parent.getResolvedInstance( subData ) );
+								if ( data.getState() == State.INITIALIZED ) {
+									return voidFuture();
+								}
+								return doCreateCompositeInstance( data )
+										.thenAccept( v -> EMBEDDED_LOAD_LOGGER.debugf(
+												"Created composite instance [%s]",
+												getNavigablePath()
+										) );
+							} );
+		}
+
+		return doCreateCompositeInstance( data )
+				.thenAccept( v -> EMBEDDED_LOAD_LOGGER.debugf( "Created composite instance [%s]", getNavigablePath() ) );
+
+	}
+
+	private CompletionStage<Void> doCreateCompositeInstance(ReactiveEmbeddableInitializerData data) {
+		if ( data.getInstance() == null ) {
+			return createCompositeInstance( data )
+					.thenAccept( data::setInstance );
+		}
 		return voidFuture();
+	}
+
+	private CompletionStage<Object> createCompositeInstance(ReactiveEmbeddableInitializerData data) {
+		if ( data.getState() == State.MISSING ) {
+			return nullFuture();
+		}
+
+		final EmbeddableInstantiator instantiator = data.getConcreteEmbeddableType() == null
+				? getInitializedPart().getEmbeddableTypeDescriptor().getRepresentationStrategy().getInstantiator()
+				: data.getConcreteEmbeddableType().getInstantiator();
+		final Object instance = instantiator.instantiate( data );
+		data.setState( State.RESOLVED );
+		return completedFuture( instance );
 	}
 
 	@Override

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/sql/results/graph/entity/internal/ReactiveEntityInitializerImpl.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/sql/results/graph/entity/internal/ReactiveEntityInitializerImpl.java
@@ -254,82 +254,63 @@ public class ReactiveEntityInitializerImpl extends EntityInitializerImpl
 		}
 		final RowProcessingState rowProcessingState = data.getRowProcessingState();
 		data.setState( State.RESOLVED );
-		return assembleId( data, rowProcessingState )
-				.thenCompose( unused -> {
-					if ( data.getState() == State.MISSING ) {
-						return voidFuture();
-					}
-					else {
-						final PersistenceContext persistenceContext = rowProcessingState
-								.getSession().getPersistenceContextInternal();
-						data.setEntityHolder( persistenceContext.claimEntityHolderIfPossible(
-								data.getEntityKey(),
-								null,
-								rowProcessingState.getJdbcValuesSourceProcessingState(),
-								this
-						) );
-
-						if ( useEmbeddedIdentifierInstanceAsEntity( data ) ) {
-							data.setEntityInstanceForNotify( rowProcessingState.getEntityId() );
-							data.setInstance( data.getEntityInstanceForNotify() );
-						}
-						else {
-							return reactiveResolveEntityInstance1( data )
-									.thenAccept( v -> {
-										if ( data.getUniqueKeyAttributePath() != null ) {
-											final SharedSessionContractImplementor session = rowProcessingState.getSession();
-											final EntityPersister concreteDescriptor = getConcreteDescriptor( data );
-											final EntityUniqueKey euk = new EntityUniqueKey(
-													concreteDescriptor.getEntityName(),
-													data.getUniqueKeyAttributePath(),
-													rowProcessingState.getEntityUniqueKey(),
-													data.getUniqueKeyPropertyTypes()[concreteDescriptor.getSubclassId()],
-													session.getFactory()
-											);
-											session.getPersistenceContextInternal().addEntity(
-													euk,
-													data.getInstance()
-											);
-										}
-										postResolveInstance( data );
-									} );
-						}
-						postResolveInstance( data );
-						return voidFuture();
-					}
-				} );
-	}
-
-	private CompletionStage<Void> assembleId(
-			ReactiveEntityInitializerData data,
-			RowProcessingState rowProcessingState) {
 		if ( data.getEntityKey() == null ) {
-			DomainResultAssembler<?> identifierAssembler = getIdentifierAssembler();
-			assert identifierAssembler != null;
-			if ( identifierAssembler instanceof ReactiveDomainResultsAssembler<?> reactiveAssembler ) {
-				return reactiveAssembler
-						.reactiveAssemble( (ReactiveRowProcessingState) rowProcessingState )
-						.thenAccept( id -> {
-							if ( id == null ) {
-								setMissing( data );
-								return ;
-							}
-							resolveEntityKey( data, id );
-						} );
-			}
-			else {
-				final Object id = identifierAssembler.assemble( rowProcessingState );
-				if ( id == null ) {
-					setMissing( data );
-					return voidFuture();
-				}
-				resolveEntityKey( data, id );
-				return voidFuture();
-			}
+			return assembleId( rowProcessingState )
+					.thenCompose( id -> {
+						if ( id == null ) {
+							setMissing( data );
+							return voidFuture();
+						}
+						resolveEntityKey( data, id );
+						return postAssembleId( rowProcessingState, data );
+					} );
 		}
-		return voidFuture();
+		return postAssembleId( rowProcessingState, data );
 	}
 
+	private CompletionStage<Void> postAssembleId(RowProcessingState rowProcessingState, ReactiveEntityInitializerData data) {
+		final PersistenceContext persistenceContext = rowProcessingState.getSession().getPersistenceContextInternal();
+		data.setEntityHolder( persistenceContext.claimEntityHolderIfPossible(
+				data.getEntityKey(),
+				null,
+				rowProcessingState.getJdbcValuesSourceProcessingState(),
+				this
+		) );
+
+		if ( useEmbeddedIdentifierInstanceAsEntity( data ) ) {
+			data.setEntityInstanceForNotify( rowProcessingState.getEntityId() );
+			data.setInstance( data.getEntityInstanceForNotify() );
+			postResolveInstance( data );
+			return voidFuture();
+		}
+
+		return reactiveResolveEntityInstance1( data )
+				.thenAccept( v -> {
+					if ( data.getUniqueKeyAttributePath() != null ) {
+						final SharedSessionContractImplementor session = rowProcessingState.getSession();
+						final EntityPersister concreteDescriptor = getConcreteDescriptor( data );
+						final EntityUniqueKey euk = new EntityUniqueKey(
+								concreteDescriptor.getEntityName(),
+								data.getUniqueKeyAttributePath(),
+								rowProcessingState.getEntityUniqueKey(),
+								data.getUniqueKeyPropertyTypes()[concreteDescriptor.getSubclassId()],
+								session.getFactory()
+						);
+						session.getPersistenceContextInternal().addEntity( euk, data.getInstance() );
+					}
+					postResolveInstance( data );
+			} );
+	}
+
+	private CompletionStage<?> assembleId(RowProcessingState rowProcessingState) {
+		final DomainResultAssembler<?> identifierAssembler = getIdentifierAssembler();
+		assert identifierAssembler != null;
+		return identifierAssembler instanceof ReactiveDomainResultsAssembler<?> reactiveAssembler
+				? reactiveAssembler.reactiveAssemble( (ReactiveRowProcessingState) rowProcessingState )
+				: completedFuture( identifierAssembler.assemble( rowProcessingState ) );
+	}
+
+	// We could move this method in ORM
 	private void postResolveInstance(ReactiveEntityInitializerData data) {
 		if ( data.getInstance() != null ) {
 			upgradeLockMode( data );

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/sql/results/graph/entity/internal/ReactiveEntityInitializerImpl.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/sql/results/graph/entity/internal/ReactiveEntityInitializerImpl.java
@@ -26,6 +26,7 @@ import org.hibernate.proxy.LazyInitializer;
 import org.hibernate.proxy.map.MapProxy;
 import org.hibernate.reactive.session.ReactiveQueryProducer;
 import org.hibernate.reactive.sql.exec.spi.ReactiveRowProcessingState;
+import org.hibernate.reactive.sql.results.graph.ReactiveDomainResultsAssembler;
 import org.hibernate.reactive.sql.results.graph.ReactiveInitializer;
 import org.hibernate.sql.results.graph.AssemblerCreationState;
 import org.hibernate.sql.results.graph.DomainResult;
@@ -253,47 +254,79 @@ public class ReactiveEntityInitializerImpl extends EntityInitializerImpl
 		}
 		final RowProcessingState rowProcessingState = data.getRowProcessingState();
 		data.setState( State.RESOLVED );
-		if ( data.getEntityKey() == null ) {
-			assert getIdentifierAssembler() != null;
- 			final Object id = getIdentifierAssembler().assemble( rowProcessingState );
-			if ( id == null ) {
-				setMissing( data );
-				return voidFuture();
-			}
-			resolveEntityKey( data, id );
-		}
-		final PersistenceContext persistenceContext = rowProcessingState.getSession()
-				.getPersistenceContextInternal();
-		data.setEntityHolder( persistenceContext.claimEntityHolderIfPossible(
-				data.getEntityKey(),
-				null,
-				rowProcessingState.getJdbcValuesSourceProcessingState(),
-				this
-		) );
+		return assembleId( data, rowProcessingState )
+				.thenCompose( unused -> {
+					if ( data.getState() == State.MISSING ) {
+						return voidFuture();
+					}
+					else {
+						final PersistenceContext persistenceContext = rowProcessingState
+								.getSession().getPersistenceContextInternal();
+						data.setEntityHolder( persistenceContext.claimEntityHolderIfPossible(
+								data.getEntityKey(),
+								null,
+								rowProcessingState.getJdbcValuesSourceProcessingState(),
+								this
+						) );
 
-		if ( useEmbeddedIdentifierInstanceAsEntity( data ) ) {
-			data.setEntityInstanceForNotify( rowProcessingState.getEntityId() );
-			data.setInstance( data.getEntityInstanceForNotify() );
-		}
-		else {
-			return reactiveResolveEntityInstance1( data )
-					.thenAccept( v -> {
-						if ( data.getUniqueKeyAttributePath() != null ) {
-							final SharedSessionContractImplementor session = rowProcessingState.getSession();
-							final EntityPersister concreteDescriptor = getConcreteDescriptor( data );
-							final EntityUniqueKey euk = new EntityUniqueKey(
-									concreteDescriptor.getEntityName(),
-									data.getUniqueKeyAttributePath(),
-									rowProcessingState.getEntityUniqueKey(),
-									data.getUniqueKeyPropertyTypes()[concreteDescriptor.getSubclassId()],
-									session.getFactory()
-							);
-							session.getPersistenceContextInternal().addEntity( euk, data.getInstance() );
+						if ( useEmbeddedIdentifierInstanceAsEntity( data ) ) {
+							data.setEntityInstanceForNotify( rowProcessingState.getEntityId() );
+							data.setInstance( data.getEntityInstanceForNotify() );
+						}
+						else {
+							return reactiveResolveEntityInstance1( data )
+									.thenAccept( v -> {
+										if ( data.getUniqueKeyAttributePath() != null ) {
+											final SharedSessionContractImplementor session = rowProcessingState.getSession();
+											final EntityPersister concreteDescriptor = getConcreteDescriptor( data );
+											final EntityUniqueKey euk = new EntityUniqueKey(
+													concreteDescriptor.getEntityName(),
+													data.getUniqueKeyAttributePath(),
+													rowProcessingState.getEntityUniqueKey(),
+													data.getUniqueKeyPropertyTypes()[concreteDescriptor.getSubclassId()],
+													session.getFactory()
+											);
+											session.getPersistenceContextInternal().addEntity(
+													euk,
+													data.getInstance()
+											);
+										}
+										postResolveInstance( data );
+									} );
 						}
 						postResolveInstance( data );
-					} );
+						return voidFuture();
+					}
+				} );
+	}
+
+	private CompletionStage<Void> assembleId(
+			ReactiveEntityInitializerData data,
+			RowProcessingState rowProcessingState) {
+		if ( data.getEntityKey() == null ) {
+			DomainResultAssembler<?> identifierAssembler = getIdentifierAssembler();
+			assert identifierAssembler != null;
+			if ( identifierAssembler instanceof ReactiveDomainResultsAssembler<?> reactiveAssembler ) {
+				return reactiveAssembler
+						.reactiveAssemble( (ReactiveRowProcessingState) rowProcessingState )
+						.thenAccept( id -> {
+							if ( id == null ) {
+								setMissing( data );
+								return ;
+							}
+							resolveEntityKey( data, id );
+						} );
+			}
+			else {
+				final Object id = identifierAssembler.assemble( rowProcessingState );
+				if ( id == null ) {
+					setMissing( data );
+					return voidFuture();
+				}
+				resolveEntityKey( data, id );
+				return voidFuture();
+			}
 		}
-		postResolveInstance( data );
 		return voidFuture();
 	}
 

--- a/hibernate-reactive-core/src/test/java/org/hibernate/reactive/EmbeddedIdWithOneToOneTest.java
+++ b/hibernate-reactive-core/src/test/java/org/hibernate/reactive/EmbeddedIdWithOneToOneTest.java
@@ -1,0 +1,131 @@
+/* Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright: Red Hat Inc. and Hibernate Authors
+ */
+package org.hibernate.reactive;
+
+import org.junit.jupiter.api.Test;
+
+import io.vertx.junit5.VertxTestContext;
+import jakarta.persistence.Embeddable;
+import jakarta.persistence.EmbeddedId;
+import jakarta.persistence.Entity;
+import jakarta.persistence.FetchType;
+import jakarta.persistence.Id;
+import jakarta.persistence.JoinColumn;
+import jakarta.persistence.OneToOne;
+import java.util.Collection;
+import java.util.List;
+import java.util.Objects;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class EmbeddedIdWithOneToOneTest extends BaseReactiveTest {
+
+	@Override
+	protected Collection<Class<?>> annotatedEntities() {
+		return List.of( FooEntity.class, BarEntity.class );
+	}
+
+	@Test
+	public void test(VertxTestContext context) {
+		BarEntity barEntity = new BarEntity( "1" );
+		FooId fooId = new FooId( barEntity );
+		FooEntity entity = new FooEntity( fooId );
+
+		test(
+				context, getMutinySessionFactory()
+						.withTransaction( s -> s.persistAll( barEntity, entity ) )
+						.chain( () -> getMutinySessionFactory()
+								.withTransaction( s -> s.find( FooEntity.class, fooId ) )
+						)
+						.invoke( result -> {
+							assertThat( result.getId() ).isEqualTo( fooId );
+							assertThat( result.getId().getIdEntity() ).isEqualTo( fooId.getIdEntity() );
+							assertThat( result.getId().getIdEntity().getId() ).isEqualTo( fooId.getIdEntity().getId() );
+						} )
+		);
+	}
+
+	@Entity(name = "bar")
+	public static class BarEntity {
+
+		@Id
+		private String id;
+
+		public BarEntity() {
+		}
+
+		public BarEntity(String id) {
+			this.id = id;
+		}
+
+		public String getId() {
+			return id;
+		}
+
+		public void setId(String id) {
+			this.id = id;
+		}
+	}
+
+	@Entity(name = "foo")
+	public static class FooEntity {
+
+		@EmbeddedId
+		private FooId id;
+
+		public FooEntity() {
+		}
+
+		public FooEntity(FooId id) {
+			this.id = id;
+		}
+
+		public FooId getId() {
+			return id;
+		}
+
+		public void setId(FooId id) {
+			this.id = id;
+		}
+	}
+
+	@Embeddable
+	public static class FooId {
+
+		@OneToOne(fetch = FetchType.EAGER)
+		@JoinColumn(name = "id", nullable = false)
+		private BarEntity barEntity;
+
+		public FooId() {
+		}
+
+		public FooId(BarEntity barEntity) {
+			this.barEntity = barEntity;
+		}
+
+		public BarEntity getIdEntity() {
+			return barEntity;
+		}
+
+		public void setIdEntity(BarEntity barEntity) {
+			this.barEntity = barEntity;
+		}
+
+		@Override
+		public boolean equals(Object o) {
+			if ( o == null || getClass() != o.getClass() ) {
+				return false;
+			}
+			FooId fooId = (FooId) o;
+			return Objects.equals( barEntity, fooId.barEntity );
+		}
+
+		@Override
+		public int hashCode() {
+			return Objects.hashCode( barEntity );
+		}
+	}
+}

